### PR TITLE
[shopsys] export js translations

### PR DESCRIPTION
**Pull request by pesektomas**
| Q             | A
| ------------- | ---
|Description, reason for the PR| We started using ES6+ syntax for javascripts in #1545, but current JsCompiler doesn't understand newest ES syntax. We had to change approach to exporting translations from javascripts - we created js command to export all used javascript messages to temporary json file and this json file we load in JsFileExtractor.
|New feature| Yes 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| No 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: packages/framework/src/Component/Translation/JsFileExtractor.php [hunk 1]
```diff
@@ -6,63 +6,48 @@
...
-    }
-
     /**
      * @param \SplFileInfo $file
      * @param \JMS\TranslationBundle\Model\MessageCatalogue $catalogue
      */
     public function visitFile(SplFileInfo $file, MessageCatalogue $catalogue)
     {
-        if (substr($file, -3) !== '.js') {
+        if (substr($file, -strlen(static::DUMP_FILE)) !== static::DUMP_FILE) {
```
Comment (grossmannmartin) [inline packages/framework/src/Component/Translation/JsFileExtractor.php]: This condition is really hard to understand to me and I'm not really sure what this checking.
Can you please extract it into a separate method and name accordingly?

File: packages/framework/src/Component/Translation/JsFileExtractor.php [hunk 2]
```diff
@@ -6,63 +6,48 @@
...
-            throw new \Shopsys\FrameworkBundle\Component\Translation\Exception\ExtractionException(
-                $ex->getMessage() . "\n" . 'in file ' . $this->file->getRealPath()
-            );
-        } catch (\Shopsys\FrameworkBundle\Component\Javascript\Parser\Exception\JsParserException $ex) {
-            throw new \Shopsys\FrameworkBundle\Component\Translation\Exception\ExtractionException(
-                $ex->getMessage() . ' in file ' . $this->file->getRealPath()
+        foreach ($translationsDump as $translation) {
+            $message = new Message(
+                $translation['id'],
+                array_key_exists('domain', $translation) ? $translation['domain'] : static::DEFAULT_MESSAGE_DOMAIN
```
Comment (grossmannmartin) [inline packages/framework/src/Component/Translation/JsFileExtractor.php]: _nitpick_ 🤓 

You can use `$translation['domain'] ?? static::DEFAULT_MESSAGE_DOMAIN`

Comment (grossmannmartin) [reply packages/framework/src/Component/Translation/JsFileExtractor.php]: also, the domain key is the only one that is optional?

File: project-base/assets/js/bin/trans.js [hunk 3]
```diff
@@ -1,8 +1,18 @@
 #!/usr/bin/env node
 
 const processTrans = require('../commands/translations/process');
-const dirWithJsFiles = './assets/js/';
-const dirWithTranslations = './translations/';
-const outputDirForExportedTranslations = dirWithJsFiles;
+const dirWithTranslations = './translations/*.po';
+const outputDirForExportedTranslations = './assets/js/';
```
Comment (grossmannmartin) [inline project-base/assets/js/bin/trans.js]: current dir (./) is the root of the project? Is it somehow dependent on where I run the command? (is it even possible?)

File: project-base/assets/js/bin/transDump.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    const keyValueParameter = parameter.split('=');
+    if (keyValueParameter[0] === 'source-dir') {
+        return keyValueParameter[1];
+    }
+}).filter(item => item !== undefined);
+
+if (dirsWithJsFiles.length === 0) {
+    dirsWithJsFiles.push(defaultDirWithJsFiles + '**/*.js');
+}
+
```
Comment (grossmannmartin) [inline project-base/assets/js/bin/transDump.js]: The (almost) same code is in trans.js. Does it make sense to extract into a separate helper function?

File: project-base/assets/js/commands/translations/findAndSaveTranslations.js [hunk 5]
```diff
@@ -39,21 +38,25 @@ function saveTranslations (translations, dirWithJsFiles, dirWithTranslations, ou
...
-            });
+                    allTranslations[translatedObject.lang] = allTranslations[translatedObject.lang].concat(translatedObject.translated);
+                });
 
             fs.writeFile(outputDirForExportedTranslations + 'translations.json', JSON.stringify(allTranslations), (writeErr) => {
                 if (writeErr) {
                     return console.log(writeErr);
                 }
+
+                return console.log('Translations was save in ' + outputDirForExportedTranslations + 'translations.json');
```
Comment (grossmannmartin) [inline project-base/assets/js/commands/translations/findAndSaveTranslations.js]:
```suggestion
                return console.log('Translations were saved to ' + outputDirForExportedTranslations + 'translations.json');
```

File: project-base/assets/js/commands/translations/findLang.test.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import findLang from './findLang';
+
+test.each([
+    ['', undefined],
+    ['lang', undefined],
+    ['lang.po', 'lang'],
+    ['lang.en.po', 'en'],
+    ['lang.js', 'lang'],
```
Comment (grossmannmartin) [inline project-base/assets/js/commands/translations/findLang.test.js]: method name tells me it finding language (presumably from filename). Now I'm confused, is it ok to return `lang` from `lang.js`? Can you please explain the reason to me?

File: project-base/package.json [hunk 7]
```diff
@@ -32,21 +38,23 @@
...
         "stylelint": "^11.1.1",
         "time-grunt": "^1.4.0",
-        "webpack-notifier": "^1.6.0",
-        "pofile": "^1.1.0"
+        "webpack-notifier": "^1.6.0"
     },
     "dependencies": {
         "@shopsys/framework": "9.0.0-dev.1",
         "codemirror": "^5.49.2",
+        "glob": "^7.1.6",
```
Comment (grossmannmartin) [inline project-base/package.json]: glob is not just a dev dependency?

File: build.xml [hunk 8]
```diff
@@ -182,6 +182,16 @@
         <phingcall target="eslint-fix"/>
     </target>
 
+    <target name="npm-export-translations" description="Installs node modules dependencies." hidden="true">
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply build.xml]: My mistake, thank you for point. 98b1643fbb776bbf637518514ccd732742ca500c

File: packages/framework/build.xml [hunk 9]
```diff
@@ -1100,7 +1115,7 @@
         </if>
     </target>
 
-    <target name="tests-unit" depends="production-protection,clean" description="Runs unit tests.">
+    <target name="tests-unit" depends="tests-unit-jest,production-protection,clean" description="Runs unit tests.">
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply packages/framework/build.xml]: I don't know why I thought that sorting is reverse. Thanks for point.8e4c21b449ef5d6c2b1b0632edea9bb22b907195

File: packages/framework/build.xml [hunk 10]
```diff
@@ -673,6 +678,16 @@
         </exec>
     </target>
 
+    <target name="npm-translations-dump" description="Dumped translations from javascript's files." hidden="true">
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply packages/framework/build.xml]: It's the some as you wrote. Destination of `*.po` files is directly in `assets/js/bin/trans.js`.

File: build.xml [hunk 11]
```diff
@@ -194,6 +204,24 @@
         </exec>
     </target>
 
+    <target name="npm-translations-dump" description="Dumped translations from javascript's files." hidden="true">
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply build.xml]: Thanks for testing, I don't understand why it worked on my instance. A fixed it 3c87b075b3348da6293a91e67ffbcaba6a265dad

File: packages/framework/src/Component/Translation/JsFileExtractor.php [hunk 12]
```diff
@@ -6,63 +6,48 @@
...
 
 class JsFileExtractor implements FileVisitorInterface
 {
-    /**
-     * @var \SplFileInfo
-     */
-    protected $file;
+    public const DUMP_FILE = 'translationsDump.json';
+
+    public const DEFAULT_MESSAGE_DOMAIN = 'messages';
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply packages/framework/src/Component/Translation/JsFileExtractor.php]: I thought about that someone would like change destinations and files with translations (in right - I don't know why). It is wrong idea?

File: packages/framework/src/Component/Translation/JsFileExtractor.php (hunk 1 above)
Comment (pesektomas) [reply packages/framework/src/Component/Translation/JsFileExtractor.php]: c192df736b379e76439b97301e3593197d8691c7

File: packages/framework/src/Component/Translation/JsFileExtractor.php (hunk 2 above)
Comment (pesektomas) [reply packages/framework/src/Component/Translation/JsFileExtractor.php]: In this case is domain only one optional key. Dumped object has one more optional key - locale, but it is not use in this class. Nitpick: 3a0f789139278d5e60e08320f03565a40150ddc7

File: project-base/assets/js/bin/trans.js (hunk 3 above)
Comment (pesektomas) [reply project-base/assets/js/bin/trans.js]: You have to run command in same path where is package.json. Without package.json the command (and other commands) don't work.

File: project-base/assets/js/bin/transDump.js (hunk 4 above)
Comment (pesektomas) [reply project-base/assets/js/bin/transDump.js]: d492c98d79ea489dc00c94f8dc2b05ea3c13ebfa

File: project-base/assets/js/commands/translations/findAndSaveTranslations.js (hunk 5 above)
Comment (pesektomas) [reply project-base/assets/js/commands/translations/findAndSaveTranslations.js]: f43a15a64332f6d2edaa138bd3a26394450e2f16

File: project-base/package.json (hunk 7 above)
Comment (pesektomas) [reply project-base/package.json]: a690fce572837c5638850736c313a2e3924eed01

File: project-base/assets/js/commands/translations/findLang.test.js (hunk 6 above)
Comment (pesektomas) [reply project-base/assets/js/commands/translations/findLang.test.js]: 5a34a254c46c71fdf33b29d212d0a09b896ddae0

**Pull request merged**